7659 Bug-fixes i periode og vedtak

ragnarwestad opened the pull request:
https://jira.adeo.no/browse/MELOSYS-7659

Fikser disse bugs nevnt i JIRA kommentar:
- "Datovalidering for ugyldig dato mangler. Nå feiler det med 2 ulike kontekster. Se steg 14 i [testskriptet](https://jira.adeo.no/secure/Tests.jspa#/testPlayer/testExecution/MELOSYS-E745) [Chak Kwan Hui](https://jira.adeo.no/secure/ViewProfile.jspa?name=H173331) 

- Radiobuttons "ja"/"nei" på vedtakssteget når man kan valgt art. 11(5) skal være sidestilt, se oppdatert skisse etter en prat med [Cathrin Engen](https://jira.adeo.no/secure/ViewProfile.jspa?name=E162585) idag   Og det skal ikke være default "nei", men saksbehandler må ta et aktivt valg her og velge. Så det må være en validering på at saksbehandler har valgt ja/nei"

VurderingPeriode-steget (EU/EØS)

  - Fikset datovalidering: Ugyldig input som "s" viser nå riktig feilmelding "Skriv inn en gyldig dato"
  - Fikset schema-rekkefølge: Endret til .required().erGyldigDato() slik at tomme felt gir "Må fylles ut" og ugyldige datoer gir "Skriv inn en gyldig dato"
  - Forenklet useEffect-håndtering:
    - Bruker null som fallback i formatterDatoTilISO() for å hindre at "Invalid date" (truthy) trigger oppdatering
    - Fjernet useEffect som kalte reset() basert på Redux-props - flyttet logikken til defaultValues
    - Redusert fra 4 til 3 useEffect-er
  - Forbedret visning av lovvalgsperiode: Viser søknadsperiode som fallback ved ugyldig dato-input

  Vedtakssteget (EU/EØS art. 11(5)):

  - Radiogruppe for "Informer utenlandsk trygdemyndighet" er nå horisontal (HStack)
  - Fjernet lokal vedtakPending state, bruker nå Redux-selector vedtakSelectors.ErPendingSelector
  - Forenklet onSubmit (fjernet manuell pending-håndtering)
  - Fjernet default-verdi for informerUtenlandskTrygdemyndighet (nå null i stedet for true/false)
  - Fikset at "Fatt vedtak"-knappen er disabled mens vedtak er pending
  - Endret useEffect-dependency fra formIsValid til formValues?.vedtakstype

ragnarwestad committed:
7659  Forbedret datovalidering i vurderingPeriode-steget

  Validering:
  - Endret rekkefølge på yup-validatorer (required → erGyldigDato → erInnenforSoknadsperioden)
  - La til forhindreAutoUtfylling og brukInternValidering props på datofeltene

  Datovelger-forbedringer:
  - Fikset onDateChange til å ikke overskrive brukerens input med undefined
  - La til onChange(currentValue) i handleOnBlur for korte inputs
  - Eksponerte brukInternValidering prop gjennom Forms.Datovelger wrapper

  Tester:
  - La til 10 nye enhetstester for datovelger
  - Dekker brukInternValidering, forhindreAutoUtfylling og edge cases
  - Alle 1728 tester passerer

ragnarwestad committed:
7659 Fikset horisontal radiogruppe og fjernet default-verdi for art. 11(5) vedtak

  - Bruker Nav.HStack for horisontal layout av ja/nei radioknapper
  - Fjernet default "nei"-verdi slik at saksbehandler må ta aktivt valg
  - Validering krever at ja/nei er valgt før man kan fortsette

ragnarwestad committed:
7659 Forbedret vedtakssteget

  Endringer:
  - Horisontale radioknapper for "Skal utenlandsk trygdemyndighet informeres?" (for art. 11(5))
    - Fjernet default "nei" - saksbehandler må gjøre aktivt valg
    - Fikset knapp-flimring på "Fatt vedtak" knappen ved klikk på radioknapper (useEffect-dependency)

  - Dobbelklikk-beskyttelse på "Fatt vedtak" via Redux pending-state
    - Erstattet lokal useState med vedtakSelectors.ErPendingSelector
    - Refaktorert onSubmit til async/await

ragnarwestad committed:
Merge branch 'master' into 7659-fixes-i-periode-og-vedtak

thanglyphan commented on src/sider/eu_eøs/stegKomponenter/vurderingPeriode/vurderingPeriode.tsx:
```diff
@@ -232,7 +232,14 @@ export function VurderingPeriode({
           <Nav.Row>
             <Nav.Column xs="3">
-              <Forms.Datovelger label="Startdato" name="fomDato" control={control} readOnly={!redigerbart} />
+              <Forms.Datovelger
```
Jeg antar det er test 14 denne skal fikse? https://jira.adeo.no/secure/Tests.jspa#/testPlayer/testExecution/MELOSYS-E745

I schema for vurderingPerioder (FTRL) som også bruker datovalidering, fant jeg: 

```
          .test(utenforSøknadsperioden)
          .test(åpenTomNårIkkeSistePeriodeTest)
          .test(tomDatoPåkrevd),
```

Vil dette hjelpe om det lå inne i schema for vurderingPeriode - EØS, eller hva er grunnen til å utvide komponenten her?

ragnarwestad replied:
Det her var dessverre ikke godt forklart av meg. 

brukInternValidering:
Under rettingen av problemene med test 14 med feil feilmeldinger så jeg at mens feil input "25.12.2025s" ga riktig feilmelding "Skriv inn en gyldig dato", så ga ikke input "s" noen feilmelding i det hele tatt (tidligere kom "Må fylles ut" som ikke er riktig)
Jeg fant ut at hvis vi la på brukInternValidering så får vi riktig feilmelding "Skriv inn en gyldig dato". Det aktiverer valideringen i DatePicker som håndterer tekstinput.

Schema-validering kan ikke løse dette fordi feilen skjer i synkroniseringen mellom DatePicker og react-hook-form, ikke i valideringslogikken. Skjemaet mottar aldri den ugyldige verdien - den blir overskrevet til tom streng før skjemaet kjører.
  1. Ved ren bokstav-input ("s") får skjemaet tom streng pga. DatePicker-synkronisering
  2. Ved "dato + bokstav" ("01.12.2025s") auto-korrigerer DatePicker før feilmeldingen vises

forhindreAutoUtfylling:
Denne stopper auto-utfylling fra feks "01.01" til "01.01.2025" som gir feilmeldingen "Utenfor søknadsperioden". Eller auto-utfylling til helt feil verdier "1" -> "01.01.1001".
Men siden denne ikke løser noe av test 14 og at muligens noen brukere får noe ut av den selv om funksjonaliteten ser rar ut, så fjerner jeg den igjen.

thanglyphan replied:
Jeg har testet litt og sjekket koden, når jeg endrer denne useEffecten her: 

```
// Oppdater lovvalgsperioden når brukeren endrer datoer eller checkbox
  useEffect(() => {
    if (redigerbart) {
      const fomDato =
        formValues.forkortLovvalgsperiode && formValues.fomDato
          ? Utils.dato.formatterDatoTilISO(formValues.fomDato)
          : mottatteOpplysningerFom;
      const tomDato =
        formValues.forkortLovvalgsperiode && formValues.tomDato
          ? Utils.dato.formatterDatoTilISO(formValues.tomDato)
          : mottatteOpplysningerTom;

      // Bare oppdater hvis vi har gyldige datoer
      if (fomDato && tomDato) {
        oppdaterData(
          lagLovvalgsperiode({
            fomDato,
            tomDato,
          }),
        );
      }
    }
  }, [formValues.forkortLovvalgsperiode, formValues.fomDato, formValues.tomDato, redigerbart]);
```

til å bli:

```
  // Oppdater lovvalgsperioden når brukeren endrer datoer eller checkbox
  useEffect(() => {
    if (redigerbart) {
      const fomDato =
        formValues.forkortLovvalgsperiode && formValues.fomDato && Utils.dato.formatterDatoTilISO(formValues.fomDato);
      const tomDato = formValues.forkortLovvalgsperiode && Utils.dato.formatterDatoTilISO(formValues.tomDato);
      // Bare oppdater hvis vi har gyldige datoer
      if (fomDato && tomDato) {
        oppdaterData(
          lagLovvalgsperiode({
            fomDato,
            tomDato,
          }),
        );
      }
    }
  }, [formValues.forkortLovvalgsperiode, formValues.fomDato, formValues.tomDato, redigerbart]);
```
gir "s" skriv inn riktig dato. formatterDatoTilISO gir invalid date når det er "s", det er jo riktig. Hvorfor skal vi fallbacke til `mottatteOpplysningerTom` hvis det er invalid?

thanglyphan replied:
```
  // Oppdater form-verdiene hvis det finnes en eksisterende forkortet lovvalgsperiode
  useEffect(() => {
    if (
      lovvalgsFomDato &&
      lovvalgsTomDato &&
      (lovvalgsFomDato !== mottatteOpplysningerFom || lovvalgsTomDato !== mottatteOpplysningerTom)
    ) {
      // Det finnes en eksisterende forkortet periode - oppdater form
      reset({
        forkortLovvalgsperiode: true,
        fomDato: Utils.dato.formatterDatoTilNorsk(lovvalgsFomDato),
        tomDato: Utils.dato.formatterDatoTilNorsk(lovvalgsTomDato),
        lovvalgsbestemmelse: lovvalgsbestemmelseSomSkalVises || undefined,
      });
    }
  }, [lovvalgsFomDato, lovvalgsTomDato]);
```

Denne useEffecten her har også noe å si i valideringen. Invalid date er en truthy verdi, så den går inn i if og oppdater form. Det er hvertfall slik i master nå. 

Tror problemet med datofelter kan fikses på ved å forenkle/fjerne mye av useEffect. Som vi har snakket om på teknisk møte før, hold bruken av useEffect til minimum, det valideres feil pga. det i denne oppgaven.

ragnarwestad replied:
Det høres ut som en god idè, jeg tar en runde på forenkling av useEffects

ragnarwestad replied:
Hovedproblemet med håndtering av input "s" var at vi brukte formatterDatoTilISO("s") som returnere "Invalid date" som er truthy, så endret den til formatterDatoTilISO("s", null) så blir den falsy og feilmeldingen blir riktig. Og da uten bruk av brukInternValidering.

Fjernet også den useEffecten du peker på og flyttet den essensielle logikken fra den inn i defaultValues -  den sjekker harForkortetPeriode og setter riktige verdier med én gang.

thanglyphan replied:
Høres bra ut

ragnarwestad committed:
Fjernet forhindreAutoUtfylling, siden den ikke retter

ragnarwestad committed:
7659 Fix ugyldig dato-input trigger ikke lenger reset av skjema

  Endret formatterDatoTilISO() til å bruke null som fallback i stedet
  for "Invalid date". Dette hindrer at oppdaterData kalles med ugyldige
  datoer, som tidligere trigget reset() og overskrev form-verdiene.

  Lagt til validering av datoer før de vises i lovvalgsperiode-visningen,
  slik at søknadsperioden vises som fallback ved ugyldig input.

  Fjernet brukInternValidering da schema-validering nå fungerer korrekt.

ragnarwestad committed:
7659 Forbedret håndtering av ugyldig dato-input i useEffect

  - Forenklet useEffect for oppdaterData: Fjernet unødvendig
    formValues.fomDato/tomDato-sjekk, bruker kun forkortLovvalgsperiode
    for å avgjøre om brukerens input skal brukes
  - Lagt til "Invalid date"-sjekk i useEffect for reset() for å hindre
    at form resettes med ugyldige verdier fra Redux

ragnarwestad committed:
7659 Fjernet useEffect som kalte reset() ved endring av lovvalgsperiode

  Flyttet logikken for initialisering av forkortet periode til defaultValues
  i useForm. Dette fjerner syklusen der brukerens input trigget reset() via
  Redux, som overskrev form-verdiene.

  Redusert fra 4 til 3 useEffect-er i komponenten.

thanglyphan commented on src/felleskomponenter/forms/datovelger.tsx:
```diff
@@ -75,6 +78,7 @@ const Datovelger = forwardRef<HTMLInputElement, DatovelgerProps>(
             maxDate={rest.maxDate}
             forhindreAutoUtfylling={rest.forhindreAutoUtfylling}
             laasAar={rest.laasAar}
+            brukInternValidering={rest.brukInternValidering}
```
Trenger kanskje ikke disse endringene lenger?

ragnarwestad replied:
Enig, revertert

thanglyphan commented on src/felleskomponenter/datovelger/datovelger.test.tsx:
```diff
@@ -208,4 +274,87 @@ describe("Datovelger", () => {
     });
   });
+
+  describe("brukInternValidering", () => {
```
Fjerne denne testen siden intervalidering ikke brukes lengre?

ragnarwestad replied:
Enig, revertert

thanglyphan commented on src/felleskomponenter/datovelger/datovelger.tsx:
```diff
@@ -45,8 +45,11 @@ function Datovelger({
-    onDateChange: (nyValgtDatoFraDatePicker?: Date) =>
-      onChange(Utils.dato.formatterDatoTilNorsk(nyValgtDatoFraDatePicker, false, undefined)),
+    onDateChange: (nyValgtDatoFraDatePicker?: Date) => {
+      if (nyValgtDatoFraDatePicker && !isNaN(nyValgtDatoFraDatePicker.getTime())) {
```
Er det ikke forms.datovelger som er brukt i vurderingperiode fila, hvorfor trenger vi denne endringen?

ragnarwestad replied:
Revertert

thanglyphan commented on src/felleskomponenter/datovelger/datovelger.tsx:
```diff
@@ -116,12 +119,12 @@ function Datovelger({
       if (forhindreAutoUtfylling && !harLagtTilAar) {
         const renDato = valueToFormat.replace(/[-./]/g, "");
         if (renDato.length < 6) {
+          onChange(currentValue);
```
Trenger vi endringer i denne fila? Bare spør, i mitt hode trenger vi ikke det, men kan hende jeg overser noe

ragnarwestad replied:
Revertert

thanglyphan commented on src/felleskomponenter/datovelger/datovelger.test.tsx:
```diff
@@ -85,6 +85,35 @@ describe("Datovelger", () => {
 
+    it("sender rå verdi til onChange ved onBlur når input er for kort", () => {
+      const onChange = vi.fn();
+      render(<Datovelger {...props} onChange={onChange} forhindreAutoUtfylling={true} />);
```
Nå rendres ikke lengre datovelger med forhindreAutoUtfyllin? Eller?

ragnarwestad replied:
Ser ut til å fungere hos meg. 

Vent... jeg misforstod kanskje :-) Nei, den gjør ikke det, og jeg har revertert test-fila.

thanglyphan reviewed (changes requested):
Fant et par småting, testene er vel noe utdatert nå som det ikke brukes internvalidering og forhindreAutoUtfylling. Bør speiles.

ragnarwestad committed:
7659 Revertert tidligere forsøk på fiks i datovelger-komponentene:
  - Fjernet brukInternValidering-prop fra Forms/datovelger.tsx
  - Fjernet onDateChange-validering i datovelger.tsx
  - Fjernet onChange-kall i forhindreAutoUtfylling-blokken
  - Revertert tilhørende tester

thanglyphan reviewed (approved):
Ser bra ut!
